pkg/planner/core/planbuilder.go

```diff
@@ -2385,6 +2391,37 @@ func (b *PlanBuilder) buildAnalyzeFullSamplingTask(
 		b.ctx.GetSessionVars().StmtCtx.AppendWarning(errors.NewNoStackError("The version 2 stats would ignore the INCREMENTAL keyword and do full sampling"))
 	}
 
+	isAnalyzeTable := len(as.PartitionNames) == 0
+	allSpecialGlobalIndex := true
+	if as.IndexFlag && len(as.IndexNames) == 0 {
+		for _, idx := range tbl.TableInfo.Indices {
+			if idx.State != model.StatePublic {
+				continue
+			}
+			if !handleutil.IsSpecialGlobalIndex(idx, tbl.TableInfo) {
+				allSpecialGlobalIndex = false
+				break
+			}
+			if !isAnalyzeTable {
+				return errors.NewNoStackErrorf("Analyze global index '%s' can't work with analyze specified partitions", idx.Name.O)
+			}
+		}
+	} else {
+		for _, idxName := range as.IndexNames {
+			idx := tbl.TableInfo.FindIndexByName(idxName.L)
+			if idx == nil || idx.State != model.StatePublic {
+				return plannererrors.ErrAnalyzeMissIndex.GenWithStackByArgs(idxName.O, tbl.Name.O)
+			}
+			if !handleutil.IsSpecialGlobalIndex(idx, tbl.TableInfo) {
+				allSpecialGlobalIndex = false
+				break
+			}
+			if !isAnalyzeTable {
+				return errors.NewNoStackErrorf("Analyze global index '%s' can't work with analyze specified partitions", idx.Name.O)
+			}
+		}
+	}
+
```

Comment:
How about moving this check to a separate function?

Reply:
Done, PTAL